pkg/cri/sbserver/sandbox_run.go

```diff
@@ -159,21 +206,31 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		sandboxCreateNetworkTimer.UpdateSince(netStart)
 	}
 
-	// Save sandbox metadata to store
 	if err := sandboxInfo.AddExtension("metadata", &sandbox.Metadata); err != nil {
 		return nil, fmt.Errorf("unable to save sandbox %q to store: %w", id, err)
 	}
-
-	if _, err := c.client.SandboxStore().Create(ctx, sandboxInfo); err != nil {
-		return nil, fmt.Errorf("failed to save sandbox metadata: %w", err)
+	// Save sandbox metadata to store
+	if sandboxInfo, err = c.client.SandboxStore().Update(ctx, sandboxInfo, "extensions"); err != nil {
+		return nil, fmt.Errorf("unable to update extensions for sandbox %q: %w", id, err)
 	}
 
 	runtimeStart := time.Now()
 
 	resp, err := c.sandboxController.Start(ctx, id)
 	if err != nil {
+		sandbox.Container, _ = c.client.LoadContainer(ctx, id)
+		if resp != nil && resp.SandboxID == "" && resp.Pid == 0 && resp.CreatedAt == nil && len(resp.Labels) == 0 {
+			// if resp is a non-nil zero-value, an error occurred during cleanup
+			cleanupErr = fmt.Errorf("failed to cleanup sandbox")
+		}
```

Comment:
Should Start() return an error instead of asking the caller to check whether resp is its zero value or not?

Reply:
I struggled with this question too.  `Start()` does already return an error in the case when the sandbox couldn't be started, and `Start()` attempts to roll-back the changes made in order to setup and start the sandbox.  The challenge is that there's an additional indication needed for the case where `Start()` fails in its cleanup.  Rather than add another `error` and change the `sandbox.Controller` interface, I opted for doing this zero-value approach instead.  There's a note at the top of `Start()` now to check whether this is reasonable once `Delete()` is implemented, since an alternative would be to remove all cleanup actions from `Start()` and instead expect that the caller invoke `Delete()` explicitly if `Start()` returns an error.

I'm open to other suggestions here though if you've got some.

Reply:
If `Start()` fails, `Start()` should internally roll-back what it created itself. If `Start()` succeeds, then `Delete()` needs to be called to clean up what created by Start().

Reply:
> If Start() fails, Start() should internally roll-back what it created itself. 

This is discussing the case where `Start()` fails to roll-back what it created.

Reply:
Ok, I mean, The rollback should eventually be handled within the `SandboxController`. includes handling Sandbox 's `TaskEvents` with `EventMonitor` in the `SandboxController`. The CRI should not be aware of the `SandboxContainer`, it only uses the `SandboxController`.

Of course, as an intermediate process, the current implementation is possible. I just imagine the final state.

Reply:
Yep.  This PR is meant to port over the bugfix behavior of #5904 and enable the new tests as we continue to work on the overall Sandbox API support.  I expect that as more of the `SandboxController` and `sbserver` are implemented, some of this code will need to change (though the tests _shouldn't_ as those should help prevent behavior regression).

Reply:
yes, I appreciate your work. It takes a lot of time. 
If possible we can communicate together. I am very interested in this. :)

Reply:
And I hope not to trouble you because of the pit I wrote.

Reply:
No trouble, thanks for taking the time to look at the PR and share your feedback!